### DISCUSSION TITLE
How can I prevent the modal content from going empty while the dismiss animation is still ongoing?

### DISCUSSION THREAD

**Discussion by Jager-yoo** [pointfreeco/swift-composable-architecture#2004] [Q&A] [answered]
Hey everyone! 👋🏻
I've encountered an issue when using SwiftUI's `sheet` or `fullScreenCover` with an optional state and `IfLetStore`.
Whenever I dismiss the modal programmatically, its content immediately turns 'empty'.

This is because the `IfLetStore` re-computes its content depending on if an optional state is `nil` or non-`nil`.

For reference, please check out the following recording.
I used a [sheet example](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Examples/CaseStudies/SwiftUICaseStudies/03-Navigation-Sheet-LoadThenPresent.swift) from SwiftUICaseStudies and simply added a dismiss button using `@Environment(\.dismiss)` along with a black background for better contrast.

```swift
.sheet(
  isPresented: viewStore.binding(
    get: \.isSheetPresented,
    send: LoadThenPresent.Action.setSheet(isPresented:)
  )
) {
  IfLetStore(
    self.store.scope(
      state: \.optionalCounter,
      action: LoadThenPresent.Action.optionalCounter
    )
  ) {
    CounterView(store: $0) // This is the ⬛️ page, and turns empty when dismiss() is called
  }
}
```

https://user-images.githubusercontent.com/71127966/227717385-545152fc-d5d9-490b-b386-5f078f96994a.mp4

<br>

Does anyone have a good solution for maintaining the modal content while the dismiss animation is in progress? 🤔 
What's the best way to present modal content that has an optional state?

**Comment by tgrapperon** [comment, marked as answer]
Hey @Jager-yoo! There is a [`returningLastNonNilValue`](https://github.com/pointfreeco/swift-composable-architecture/blob/navigation-beta/Sources/ComposableArchitecture/Internal/ReturningLastNonNilValue.swift) utility in that you can use to keep presenting the last non-nil value of State. 
Please note that the upcoming navigation utilities will handle this automatically.

**Reply by Jager-yoo** [reply]
Oh wow, `returningLastNonNilValue` is so useful when presenting optional state!
I'll use this as ad hoc solution until navigation 1.0 comes out! Thanks as always. 👍🏻

```swift
.sheet(
  isPresented: viewStore.binding(
    get: \.isSheetPresented,
    send: LoadThenPresent.Action.setSheet(isPresented:)
  )
) {
  IfLetStore(
    self.store.scope(
      state: returningLastNonNilValue { $0.optionalCounter }, // ✨applied here!
      action: LoadThenPresent.Action.optionalCounter
    )
  ) {
    CounterView(store: $0)
  }
}
```

https://user-images.githubusercontent.com/71127966/227758426-9eebef77-88d0-4df2-b7a8-225db9710bf4.mp4

**Reply by Jager-yoo** [reply]
However, I'm still wondering why returningLastNonNilValue was implemented as an open function.
Could we wrap it within some extension instead?

**Reply by tgrapperon** [reply]
I can't say for sure, but my guess is that the free function is more lightweight and generic than an extension, and it makes the storage simple enough in this "surrounding" position.

**Reply by Jager-yoo** [reply]
Thank you for your help!
